server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -3969,6 +3985,20 @@ public void afterRefresh(boolean didRefresh) {
         }
     }
 
+    private class RefreshFieldHasValueListener implements ReferenceManager.RefreshListener {
+        @Override
+        public void beforeRefresh() {}
+
+        @Override
+        public void afterRefresh(boolean didRefresh) {
+            if (enableFieldHasValue) {
+                try (Engine.Searcher hasValueSearcher = getEngine().acquireSearcher("field_has_value")) {
+                    fieldInfos(FieldInfos.getMergedFieldInfos(hasValueSearcher.getIndexReader()));
```

Comment:
@original-brownbear mentioned to me that the previous map was wasting resources, and that we could just hold references to something that's already hot in the heap. I do think that merging does not do that, in that it creates a new FieldInfos class. I cannot tell how much a problem that is. Technically, we could merge once field infos are retrieved, but I am not sure how much of a difference that would make.

Reply:
This solution seemed cleaner but I understand your point. Instead of merging FieldInfos I'll keep a list and loop. This will use data already present in the heap, will avoid merging FieldInfos at each refresh and it will shift the cost (looping for each FieldInfos and for each FieldInfo) directly to the _field_caps request instead of paying it at each refresh.

Reply:
I don't know if mine was a suggestion, it was more of a point for discussion. I don't know how much of a difference it makes, perhaps @original-brownbear can comment on it.

Reply:
The resource waste in `FieldInfos` comes from duplicate strings in it. Merging doesn't hurt us here. I think we want it as it is coded here i.e. have a single merged field infos instance. That should (as far as I can see in the Lucene code) use the same objects (at least the strings) that we already reference from `SegmentReader` instances.

Reply:
Reverted back to merging. It's much cleaner :D

Reply:
sounds good to me.